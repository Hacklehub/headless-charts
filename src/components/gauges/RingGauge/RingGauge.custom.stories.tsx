/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable @typescript-eslint/no-explicit-any */
import { Meta, StoryObj } from '@storybook/react';

import RingGauge from '.';
import metrics from './sample.json';

const meta: Meta<typeof RingGauge> = {
  title: 'Gauge/RingGauge/Customized',
  component: RingGauge,
  tags: ['autodocs'],
};

export default meta;

type Story = StoryObj<typeof RingGauge>;

export const WithCustomStartAndEndAngle: Story = {
  args: {
    data: metrics.map(({ className, ...metric }: any) => metric),
    id: 'ring-chart-custom-start-end-angle',
    labelKey: 'name',
    dataKey: 'score',
    targetKey: 'target',
    startAngle: 45,
    endAngle: 180,
  },
};

/** labels at bottom position */
export const WithLabelsAtBottom: Story = {
  args: {
    ...WithCustomStartAndEndAngle.args,
    id: 'ring-chart-labels-bottom',
    startAngle: -90,
    endAngle: 180,
    labels: {
      position: 'bottom',
    },
  },
};
