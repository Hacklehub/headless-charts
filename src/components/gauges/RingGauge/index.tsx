import { PieArcDatum, arc } from 'd3-shape';
/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable @typescript-eslint/no-explicit-any */
import React, { useEffect } from 'react';
import { defaultChartClassNames, mergeTailwindClasses } from '../../../utils';
import { pointer, select, selectAll } from 'd3-selection';

// import { axisBottom } from 'd3-axis';
import { interpolateNumber } from 'd3-interpolate';
import { min } from 'd3-array';
import { scaleLinear } from 'd3-scale';
import { transition } from 'd3-transition';

export interface RingGaugeProps {
  id: string;
  className?: string;
  labelKey: string;
  targetKey: string;
  dataKey: string;
  errorKey?: string;
  labels?: {
    position?: 'top' | 'bottom';
    className?: string;
  };
  data?: any[];
  margin?: {
    top?: number;
    bottom?: number;
    left?: number;
    right?: number;
  };
  padding?: {
    arc?: number;
  };
  minRadius?: number;
  startAngle?: number;
  endAngle?: number;
  cornerRadius?: number;
  drawing?: {
    duration?: number;
    delay?: number;
  };
  tooltip?: {
    className?: string;
    html?: (d: any) => string;
  };
  classNameGauge?: string;
  classNameGaugeBg?: string;
}

const RingGauge = ({
  className,
  id,
  labelKey,
  targetKey,
  dataKey,
  data = [],
  margin = {
    top: 40,
    bottom: 40,
    left: 40,
    right: 40,
  },
  padding = {
    arc: 5,
  },
  minRadius = 10,
  cornerRadius = 2,
  drawing = { duration: 1000, delay: 0 },
  startAngle = 0,
  endAngle = 270,
  tooltip,
  classNameGaugeBg = '',
  labels = { position: 'top' },
}: RingGaugeProps) => {
  const PI = Math.PI,
    numArcs = data.length;

  const refreshChart = React.useCallback(() => {
    const svg = select(`#${id}`);

    svg.selectAll('*').remove();

    const tooltipDiv =
      tooltip && select('#tooltip').node()
        ? select('#tooltip')
        : select('body')
            .append('div')
            .attr('id', 'tooltip')
            .style('position', 'absolute')
            .style('opacity', '0')
            .attr('class', mergeTailwindClasses(tooltip?.className));

    const g = svg.append('g');

    const width = +svg.style('width').split('px')[0],
      height = +svg.style('height').split('px')[0];

    const getInnerRadius = (index: number) =>
        minRadius + (numArcs - (index + 1)) * (arcWidth + (padding.arc || 0)),
      getOuterRadius = (index: number) => getInnerRadius(index) + arcWidth;

    g.attr('transform', `translate(${width / 2},${height / 2})`);

    const innerWidth = width - (margin.left || 0) - (margin.right || 0),
      innerHeight = height - (margin.top || 0) - (margin.bottom || 0),
      chartRadius = Math.min(innerHeight, innerWidth) / 2;

    const scale = scaleLinear()
      .domain([0, 1])
      .range([0, (endAngle / 180) * PI]);

    const arcWidth =
      (chartRadius - minRadius - numArcs * (padding.arc || 5)) / numArcs;

    const arcFn = arc<PieArcDatum<any> | number>()
      .innerRadius((_d: any, i: number) => getInnerRadius(i))
      .outerRadius((_d: any, i: number) => getOuterRadius(i))
      .startAngle(((startAngle / 90) * PI) / 2)
      .endAngle((d: any) => scale(d))
      .cornerRadius(cornerRadius);

    g.append('g')
      .attr('class', `background-arcs`)
      .selectAll('path')
      .data(data)
      .enter()
      .append('path')
      .attr(
        'class',
        mergeTailwindClasses(
          ` fill-current text-gray-200 dark:text-gray-700`,
          classNameGaugeBg
        )
      )
      .attr('d', function (_d, i: number) {
        return arcFn(1, i);
      });

    transition();

    g.append('g')
      .attr('class', 'data')
      .selectAll('path')
      .data(data)
      .enter()
      .append('path')
      .attr('class', (d: any) =>
        mergeTailwindClasses('data-arc fill-current ', d.className)
      )
      .attr('d', '')
      .on('mouseenter', function (_event, d) {
        tooltipDiv.attr(
          'class',
          mergeTailwindClasses(tooltip?.className, 'tooltip')
        );
        tooltipDiv
          .style('opacity', 1)
          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
          // @ts-ignore
          .html(
            tooltip?.html
              ? tooltip.html(d)
              : `${d[labelKey]} <br/>${d[dataKey]}/${d[targetKey]}`
          );
      })
      .on('mousemove', function (event) {
        const [bX, bY] = pointer(event, select('body'));
        // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // @ts-ignore
        tooltipDiv.style('left', `${bX + 10}px`).style('top', `${bY + 10}px`);
      })
      .on('mouseleave', function () {
        tooltipDiv.style('opacity', 0);
        tooltipDiv.html('');
      })
      .transition()
      .duration(drawing?.duration || 1000)
      .delay(drawing?.delay || 0)
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      .attrTween('d', (d: any, i: number) => {
        const interpolate = interpolateNumber(
          0,
          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
          // @ts-ignore
          min([d[dataKey] / d[targetKey], 1])
        );
        return (t: any) => arcFn(interpolate(t), i);
      });

    labels &&
      g
        .append('g')
        .attr('class', 'labels')
        .selectAll('.labels')
        .data(data)
        .enter()
        .append('text')
        .attr('text-anchor', 'end')
        .attr('class', `fill-current text-xs ${labels.className || ''}`)
        .attr('x', -5)
        .attr('y', (_d, i) =>
          labels?.position === 'bottom'
            ? getInnerRadius(i) + arcWidth - 1
            : -getInnerRadius(i) - 2
        )
        .text((d: any) => d[labelKey]);
  }, [
    PI,
    classNameGaugeBg,
    data,
    dataKey,
    drawing?.delay,
    drawing?.duration,
    endAngle,
    id,
    labelKey,
    labels,
    margin.bottom,
    margin.left,
    margin.right,
    margin.top,
    minRadius,
    numArcs,
    padding.arc,
    startAngle,
    cornerRadius,
    targetKey,
    tooltip,
  ]);

  useEffect(() => {
    refreshChart();
    return () => {
      selectAll('#tooltip').remove();
    };
  }, [data, refreshChart]);
  return (
    <svg
      id={id}
      className={mergeTailwindClasses(defaultChartClassNames, className || '')}
    />
  );
};

export default RingGauge;
