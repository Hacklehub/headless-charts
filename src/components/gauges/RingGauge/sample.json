[
  {
    "name": "Target 1",
    "score": 75,
    "target": 90,
    "className": "fill-purple-700"
  },
  {
    "name": "Target 2",
    "score": 75,
    "target": 80,
    "className": "fill-purple-400"
  },
  {
    "name": "Target 3",
    "score": 70,
    "target": 55,
    "className": "fill-green-500"
  }
]
