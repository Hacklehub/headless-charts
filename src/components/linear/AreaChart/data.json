[
  {
    "year": 2012,
    "macbook": 23.22,
    "services": 12.89,
    "wearables": 10.76,
    "ipad": 30.94,
    "iphone": 78.69
  },
  {
    "year": 2013,
    "macbook": 21.48,
    "services": 16.05,
    "wearables": 10.12,
    "ipad": 31.98,
    "iphone": 91.28
  },
  {
    "year": 2014,
    "macbook": 24.08,
    "services": 18.06,
    "wearables": 8.379,
    "ipad": 30.28,
    "iphone": 101.99
  },
  {
    "year": 2015,
    "macbook": 25.47,
    "services": 19.91,
    "wearables": 10.07,
    "ipad": 23.23,
    "iphone": 155.04
  },
  {
    "year": 2016,
    "macbook": 22.83,
    "services": 24.35,
    "wearables": 11.13,
    "ipad": 20.63,
    "iphone": 136.7
  },
  {
    "year": 2017,
    "macbook": 25.57,
    "services": 32.7,
    "wearables": 12.83,
    "ipad": 18.8,
    "iphone": 139.34
  },

  {
    "year": 2018,
    "macbook": 25.2,
    "services": 39.75,
    "wearables": 17.38,
    "ipad": 18.38,
    "iphone": 164.89
  },
  {
    "year": 2019,
    "macbook": 25.74,
    "services": 46.29,
    "wearables": 24.48,
    "ipad": 21.28,
    "iphone": 142.38
  },
  {
    "year": 2020,
    "macbook": 28.62,
    "services": 53.77,
    "wearables": 30.62,
    "ipad": 23.72,
    "iphone": 137.78
  },

  {
    "year": 2021,
    "macbook": 35.19,
    "services": 68.42,
    "wearables": 38.37,
    "ipad": 31.86,
    "iphone": 191.97
  },
  {
    "year": 2022,
    "macbook": 40.18,
    "services": 78.13,
    "wearables": 41.24,
    "ipad": 29.29,
    "iphone": 205.49
  }
]
