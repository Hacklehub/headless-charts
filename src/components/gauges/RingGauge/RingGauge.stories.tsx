import { Meta } from '@storybook/react';
import RingGauge from '.';

/* eslint-disable @typescript-eslint/no-explicit-any */

const metrics = [
  {
    name: 'Target 1',
    score: 75,
    target: 90,
    className: 'text-purple-700',
  },
  {
    name: 'Target 2',
    score: 75,
    target: 80,
    className: 'text-purple-400',
  },
  {
    name: 'Target 3',
    score: 70,
    target: 55,
    className: 'text-green-500',
  },
];

export default {
  title: 'Gauge/RingGauge',
  component: RingGauge,
  tags: ['autodocs'],
} as Meta;

/** Default RingGauge Chart (Grouped). */
export const Default = {
  args: {
    data: metrics,
    id: 'ring-chart-default',
    labelKey: 'name',
    dataKey: 'score',
    targetKey: 'target',
  },
};

/** Ring gauge with tooltip */
export const WithTooltip = {
  args: {
    data: metrics,
    id: 'ring-chart-tooltip',
    labelKey: 'name',
    dataKey: 'score',
    targetKey: 'target',
    tooltip: {
      className: `bg-gray-800 text-white px-4 py-2 rounded-md shadow-md`,
    },
  },
};
