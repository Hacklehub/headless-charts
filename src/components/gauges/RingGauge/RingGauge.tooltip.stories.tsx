/* eslint-disable @typescript-eslint/no-unused-vars */

import { Meta, StoryObj } from '@storybook/react';

import RingGauge from '.';
import metrics from './sample.json';

const meta: Meta<typeof RingGauge> = {
  title: 'Gauge/RingGauge/Tooltips',
  component: RingGauge,
  tags: ['autodocs'],
};

export default meta;

type Story = StoryObj<typeof RingGauge>;

/** Ring gauge with tooltip */
export const WithTooltip: Story = {
  args: {
    data: metrics.map(({ className, ...metric }: any) => metric),
    id: 'ring-chart-tooltip',
    labelKey: 'name',
    dataKey: 'score',
    targetKey: 'target',
    tooltip: {
      className: `bg-gray-800 text-white px-4 py-2 rounded-md shadow-md`,
    },
  },
};

/** Ring gauge with custom tooltip */
export const WithCustomTooltip: Story = {
  args: {
    ...WithTooltip.args,
    id: 'ring-chart-custom-tooltip',
    tooltip: {
      className: `bg-gray-800 text-white px-4 py-2 rounded-md shadow-md`,
      html: (data: any) => `<div class="flex flex-col">
              <div class="text-lg">${data.name}</div>
              <div class="text-xs">${data.score} out of ${data.target}</div>
            </div>`,
    },
  },
};
