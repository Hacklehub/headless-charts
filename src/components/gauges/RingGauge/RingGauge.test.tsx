import { describe, expect, it } from 'vitest';
import { render, screen } from '@testing-library/react';

import RingGauge from '.';
import data from './sample.json';

describe('RingGauge', () => {
  it('renders RingGauge component for snapshot test', () => {
    render(<RingGauge
      id='ring-chart'
      data={data}
      labelKey='name'
      dataKey='score'
      targetKey='target' />);
    const ringGauge = screen.getByTestId('ring-gauge');
    expect(ringGauge).toMatchSnapshot();
  });

  it('renders RingGauge component', () => {
    render(<RingGauge
      id='ring-chart'
      data={data}
      labelKey='name'
      dataKey='score'
      targetKey='target' />);
    const ringGauge = screen.getByTestId('ring-gauge');
    expect(ringGauge).toBeInTheDocument();
  });

})
