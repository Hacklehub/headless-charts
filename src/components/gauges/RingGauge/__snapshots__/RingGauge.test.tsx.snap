// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RingGauge > renders RingGauge component for snapshot test 1`] = `
<svg
  class="chart dark:bg-gray-800 dark:text-gray-50 h-64 lg:w-4/12 md:w-6/12 text-gray-900 w-full"
  data-testid="ring-gauge"
  id="ring-chart"
>
  <g
    transform="translate(0,0)"
  >
    <g
      class="background-arcs"
    >
      <path
        class=" dark:text-gray-700 fill-current text-gray-200"
        d="M0,0Z"
      />
      <path
        class=" dark:text-gray-700 fill-current text-gray-200"
        d="M0,0Z"
      />
      <path
        class=" dark:text-gray-700 fill-current text-gray-200"
        d="M0,7.746A2,2,0,1,1,2.5,9.682A10,10,0,1,1,9.682,2.5A2,2,0,1,1,7.746,0L11.667,0Z"
      />
    </g>
    <g
      class="data"
    >
      <path
        class=" data-arc fill-purple-700"
        d="M0,0Z"
      />
      <path
        class=" data-arc fill-purple-400"
        d="M0,0Z"
      />
      <path
        class=" data-arc fill-green-500"
        d="M0,7.746A2,2,0,1,1,2.5,9.682A10,10,0,1,1,9.682,2.5A2,2,0,1,1,7.746,0L11.667,0Z"
      />
    </g>
    <g
      class="labels"
    >
      <text
        class="fill-current text-xs "
        text-anchor="end"
        x="-5"
        y="21.333333333333336"
      >
        Target 1
      </text>
      <text
        class="fill-current text-xs "
        text-anchor="end"
        x="-5"
        y="4.666666666666668"
      >
        Target 2
      </text>
      <text
        class="fill-current text-xs "
        text-anchor="end"
        x="-5"
        y="-12"
      >
        Target 3
      </text>
    </g>
  </g>
</svg>
`;
